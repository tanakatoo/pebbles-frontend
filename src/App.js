import './styles/app.css';
import NavBar from "./components/common/NavBar"
import "./styles/app.css"
import { useSelector } from 'react-redux';
import MyRoutes from "./MyRoutes"

function App() {
  //get the font from the store
  const font = useSelector(state => state.langFont.font)

  //we have to hardcode the string of the font because tailwind doesn't know what to use at build
  return (
    <div className={`App ${font == 'font-poppins' ? 'font-poppins' : 'font-notoJP'}`}>
      <NavBar />
      <MyRoutes />
    </div >
  );
}

export default App;
