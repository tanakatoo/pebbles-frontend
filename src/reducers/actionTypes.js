/* PROFILE action types */
export const LOAD_PROFILE = "loadProfile"
export const LOGOUT = "logout"
export const LOGIN = "login"

/* LANGUAGE action types */
export const CHANGE_LANG = "changeLang"
export const CHANGE_PAGE = "changePage"