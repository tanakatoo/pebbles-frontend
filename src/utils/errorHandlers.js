import { rest } from 'msw'

let loginResponse = {
    token: 'abcdefg',
    user: {
        id: 2,
        username: 'hello',
        role: 'regular',
        premium_join_date: '',
        premium_end_date: '',
        free_trial_start_date: '',
    }
}



export const errorHandlers = [

    //login
    rest.post('http://localhost:3001/auth/login', (req, res, ctx) => {
        const failLogin = {
            "error": {
                "message": "INVALID_CREDENTIALS",
                "status": 401
            }
        }
        console.log(res(ctx.json([failLogin])))
        return res(ctx.json([failLogin]));
    }),

    rest.post('http://localhost:3001/email', (req, res, ctx) => {
        console.log('MOCK error /email');
        const noData = {
            "error": {
                "message": "no data",
                "status": 400
            }
        }
        return res(ctx.json(noData));
    }),

]