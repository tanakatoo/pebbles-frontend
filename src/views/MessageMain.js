import React, { useEffect, useState } from 'react'
import Card from "../components/message/Card"
import MessageApi from '../api/message'
import ServerError from '../components/form/ServerError'
import Spinner from '../components/common/Spinner'
import { v4 as uuid } from "uuid"
import useSetToken from '../hooks/useSetToken'
import SearchBar from '../components/common/SearchBar'
import Dropdown from '../components/common/Dropdown'
import useClickOutside from '../hooks/useClickOutside'
import usePageText from "../hooks/usePageText"
import Protected from '../components/common/Protected'
import NoData from '../components/common/NoData'
import { Link } from 'react-router-dom'

function MessageMain() {
    const [data, setData] = useState(null)
    const [errors, setErrors] = useState([])
    const [token = null] = useSetToken()
    const [dropdown, setDropdown] = useState(false)
    const pageText = usePageText("messages")

    useEffect(() => {
        const getData = async () => {
            try {
                const res = await MessageApi.getAllUsersLatestMsg()
                setData(res)
            } catch (e) {
                if (e instanceof TypeError) {

                    //means server is down
                    setErrors(["UNKNOWN"])
                } else {

                    setErrors(e)
                }
            }
        }
        getData()
    }, [])

    const handleSearch = () => {
        //to be implemented later
        console.log('call to search for contact')
    }

    const handleDropdown = () => {
        dropdown ? setDropdown(false) : setDropdown(true)
    }

    const closeDropdown = () => {
        setDropdown(false)
    }
    const ref = useClickOutside(closeDropdown)

    const dropdownItems =
        [
            { text: pageText.BLOCK_CONTACTS, link: '/users/block' },
            { text: pageText.UNBLOCK_CONTACTS, link: '/users/unblock' }
        ]

    return (
        <Protected>
            <div className=' my-16 flex justify-center'>

                {errors.length > 0 && <ServerError msg={errors} />}
                {!data && errors.length === 0 ?
                    <Spinner />
                    : data.length === 0 ?
                        <div className='flex flex-col items-center'>
                            <NoData msg={pageText.NO_MSG} link="/users/dashboard" linkText={pageText.BACK_TO_DASHBOARD} />
                            <div className='text-link'><Link to="/users/unblock">{pageText.UNBLOCK_CONTACTS}</Link></div>
                        </div>
                        :
                        errors.length === 0 ?
                            <div className='container grid grid-cols-4 gap-x-4 md:grid-cols-8 lg:grid-cols-12  lg:mx-24 xl:mx-48'>
                                {/* <div className='col-span-full mb-4 mt-2'><SearchBar handleSearch={handleSearch} /></div> */}
                                <div className="flex col-span-full justify-end relative">
                                    <span className='cursor-pointer' ref={ref} onClick={handleDropdown}>
                                        {pageText.EDIT}
                                    </span>
                                    {dropdown && <Dropdown items={dropdownItems}
                                        divide={true}
                                        css="rounded shadow-dropdown top-8" />}
                                </div>
                                <div className='col-span-full'>
                                    {data.map(d => <Card data={d} key={uuid()} screen='main' />)}

                                </div>
                            </div>
                            : ''
                }

            </div>
        </Protected>
    )
}

export default MessageMain