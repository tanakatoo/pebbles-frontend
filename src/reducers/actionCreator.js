import axios from "axios"
import useLocalStorage from "../hooks/useLocalStorage"
import { LOGIN, LOGOUT } from "./actionTypes"
import AuthApi from "../api/auth";

export function actionLogin(username, password) {
    return async function (dispatch) {
        const res = await AuthApi.login(username, password)
        dispatch(actionLoggedIn(res))
    }
}

function actionLoggedIn(token) {
    return ({ type: LOGIN, token: token })
}


// export function actionLogout(token) {
//     return async function (dispatch) {
//         const [storageToken, setStorageToken, removeLocalStorage] = useLocalStorage("token")

//         //set the token state to null
//         setStorageToken(null)
//         //remove the token from the storage
//         removeLocalStorage()
//         dispatch(actionLoggedOut())
//     }
// }

// function actionLoggedOut() {
//     return ({ type: LOGOUT })
// }
