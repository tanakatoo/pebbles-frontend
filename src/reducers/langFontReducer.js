import { CHANGE_LANG, CHANGE_PAGE } from "./actionTypes"
import language from '../text/language.json';

console.log('navigator', navigator.language)
let initial_lang = "JA"
if (/^en\b/.test(navigator.language)) {
    console.log('ther eis en in the lang')
    initial_lang = "EN"
}

const INITIAL_STATE = { lang: initial_lang, font: language.FONTS[initial_lang], allFonts: language.FONTS }

const langFontReducer = (state = INITIAL_STATE, action) => {
    switch (action.type) {
        case CHANGE_LANG:
            //changes the language and sets the appropraite font and text 
            return { ...state, lang: action.lang, font: language.FONTS[action.lang] }

        default:
            return state

    }
}

export default langFontReducer
