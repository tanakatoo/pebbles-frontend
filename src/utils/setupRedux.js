import React from "react";
import { render } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
// As a basic setup, import your same slice reducers
import { persistStore, persistReducer, FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER } from 'redux-persist';
import { PersistGate } from 'redux-persist/integration/react'
import { ListSlice } from "../reducers/rootReducer";
import rootReducer from "../reducers/rootReducer"
import storage from 'redux-persist/lib/storage'
import storeSetup from "./storeSetup"
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2'

// const persistConfig = {
//     key: 'root', // Key prefix for the persisted state
//     storage, // Storage mechanism (e.g., local storage)
//     stateReconciler: autoMergeLevel2, // Specify which parts of the state to persist
//     whitelist: ['profile', 'langFont']
// };
// const persistedReducer = persistReducer(persistConfig, rootReducer);

// const persistedStore = persistStore(store)
// const root = ReactDOM.createRoot(document.getElementById('root'));


export function renderWithProviders(
    ui,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = storeSetup(preloadedState),
        ...renderOptions
    } = {}
) {
    function Wrapper({ children }) {

        return <Provider store={store}>{children}</Provider>;
    }

    // Return an object with the store and all of RTL's query functions
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}