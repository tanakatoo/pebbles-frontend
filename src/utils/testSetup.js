// import { persistedStore, store } from './setupRedux';
import { render } from '@testing-library/react'
import { Provider } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react'
import { setupTestStore } from './setupRedux';


const renderWithProviders = (ui, options = {}) => {
    const [store, persistedStore] = setupTestStore(options.preloadedState);

    const AllProviders = ({ children }) => {
        return (
            <Provider store={store}>
                <PersistGate loading={null} persistor={persistedStore}>
                    {children}
                </PersistGate>
            </Provider>
        )
    };

    return { store, ...render(ui, { wrapper: AllProviders, ...options }) };
}


// "re-export everything"
export * from '@testing-library/react';

export { renderWithProviders };