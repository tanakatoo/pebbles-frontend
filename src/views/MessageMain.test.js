import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { fireEvent, renderWithProviders } from '../utils/testSetup';
import MessageMain from './MessageMain';
import App from '../App';
import { waitFor } from '../utils/testSetup';
import { act } from 'react-dom/test-utils';
import { server } from '../utils/server'
import { errorHandlers } from '../utils/errorHandlers';
import { actionLogin } from '../reducers/actionCreator';
import { useDispatch } from 'react-redux';
import { store } from '../utils/setupRedux';
import { LOGIN } from '../reducers/actionTypes';

// Mock window.scrollTo()
window.scrollTo = jest.fn();
afterAll(() => {
    jest.clearAllMocks();
});
// let loginResponse = {
//     token: 'abcdefg',
//     user: {
//         id: 2,
//         username: 'hello',
//         role: 'regular',
//         premium_join_date: '',
//         premium_end_date: '',
//         free_trial_start_date: '',
//     }
// }

let loginResponse = {
    token: 'abcdefg',
    profile: {
        id: 2,
        username: 'hello',
        name: 'Hello World',
        email: "karmen.tanakaa@gmail.com",
        role: 'regular',
        premium_join_date: '',
        premium_end_date: '',
        free_trial_start_date: '',
        study_buddy_types: [],
        study_buddy_active: false
    }
};


beforeEach(async () => {
    window.localStorage.clear()

})

afterEach(() => {
    window.localStorage.clear()
})


describe('MessageMain', () => {
    test('renders without crashing', async () => {

        renderWithProviders(
            <MemoryRouter initialEntries={["/messages"]}>
                <App />
            </MemoryRouter>,
            {
                preloadedState: { profile: loginResponse }
            });


    });

    test('matches snapshot', async () => {

        const { asFragment } = renderWithProviders(
            <MemoryRouter initialEntries={["/messages"]}>
                <App />
            </MemoryRouter>,
            {
                preloadedState: { profile: loginResponse }
            });


        expect(asFragment()).toMatchSnapshot();


    });


    test('unread messages are orange and read are black', async () => {
        const { getByTestId, getByRole, getByLabelText, getByText } = renderWithProviders(
            <MemoryRouter initialEntries={["/messages"]}>
                <App />
            </MemoryRouter>,
            {
                preloadedState: { profile: loginResponse }
            });




        await waitFor(() => {
            expect(getByText(/june/)).toBeInTheDocument();
            expect(getByText(/ktoo/)).toBeInTheDocument();
            expect(getByText(/june/)).toHaveStyle('color: ');
            expect(getByText(/june/)).toHaveStyle('color: text-secondary-dark');
        });

    });

    // test that new conversations are first and different color

});