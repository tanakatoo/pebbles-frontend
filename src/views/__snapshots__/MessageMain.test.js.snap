// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageMain matches snapshot 1`] = `<DocumentFragment />`;
